inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp

```diff
@@ -739,7 +711,113 @@ void MKLDNNRNN::createPrimitive() {
         auto prim_desc = createPrimitiveDescriptor<lstm_forward::primitive_desc, lstm_forward::desc>();
         prim.reset(new lstm_forward(prim_desc));
     } else {
-        IE_THROW() << "Unknown cell type";
+        THROW_ERROR << "has unknown cell type.";
+    }
+}
+
+void MKLDNNRNN::prepareParams() {
+    const auto dataType = MKLDNNExtensionUtils::IEPrecisionToDataType(runtimePrecision);
+
+    const size_t B = getParentEdgesAtPort(0).front()->getMemory().GetShape().getStaticDims()[0];
+    const size_t SL = is_cell ? 1lu : getParentEdgesAtPort(0).front()->getMemory().GetShape().getStaticDims()[1];
+    Shape shapeS_4D{L, D, B, SC};
+    std::vector<DnnlBlockedMemoryDesc> inDataD, outDataD;
+
+    inDataD.reserve(S + 1);
+    outDataD.reserve(S + 1);
+
+    inDataD.emplace_back(Shape{SL, B, DC}, dataType, memory::format_tag::tnc);
+    outDataD.emplace_back(Shape{SL, B, SC}, dataType, memory::format_tag::tnc);
+
+    inDataD.emplace_back(shapeS_4D, dataType, memory::format_tag::ldnc);
+    outDataD.emplace_back(shapeS_4D, dataType, memory::format_tag::ldnc);
+
+    if (haveCellState(cell_type)) {
+        inDataD.emplace_back(shapeS_4D, memory::data_type::f32, memory::format_tag::ldnc);
+        outDataD.emplace_back(shapeS_4D, memory::data_type::f32, memory::format_tag::ldnc);
+    }
+
+    bool wFormatWasChanged = false;
+    // WA To avoid different weights layer and iter formats in FP32 case.
+    if (runtimePrecision == Precision::FP32) {
+        if (SL != 1 || B < optimalBatchSize) {
+            if (wFormat != mkldnn::memory::format_tag::ldigo) {
+                wFormat = mkldnn::memory::format_tag::ldigo;
+                wFormatWasChanged = true;
+            }
+        } else if (wFormat != mkldnn::memory::format_tag::any) {
+            wFormat = mkldnn::memory::format_tag::any;
+            wFormatWasChanged = true;
+        }
+    }
+    if (wFormatWasChanged) {
+        auto weightsDims = MKLDNNExtensionUtils::convertToDnnlDims(VectorDims{ L, D, DC, G, SC });
+        wDescs[0] = mkldnn::memory::desc(weightsDims, dataType, wFormat);
+        auto statesDims = MKLDNNExtensionUtils::convertToDnnlDims(VectorDims{ L, D, SC, G, SC });
+        wDescs[1] = mkldnn::memory::desc(statesDims, dataType, wFormat);
+    }
+
+    primitive_desc_iterator itpd;
+    const mkldnn::primitive_attr attr = mkldnn::primitive_attr();
+    if (cell_type == mkldnn::algorithm::vanilla_rnn) {
```

Comment:
762 - 817 looks similar with `createDescriptor` code. Is it possible to extract it into separate function?

Reply:
Done.